SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityScheduler.java

```diff
@@ -488,29 +489,95 @@ private int scheduleTasks(SingularityRequest request, RequestState state,
       }
 
       if (request.isRackSensitive() && configuration.isRebalanceRacksOnScaleDown()) {
-        List<SingularityTaskId> extraCleanedTasks = new ArrayList<>();
-        int numActiveRacks = rackManager.getNumActive();
-        double perRack = request.getInstancesSafe() / (double) numActiveRacks;
-
-        Multiset<String> countPerRack = HashMultiset.create();
-        for (SingularityTaskId taskId : remainingActiveTasks) {
-          countPerRack.add(taskId.getRackId());
-          LOG.info("{} - {} - {} - {}", countPerRack, perRack, extraCleanedTasks.size(), taskId);
-          if (countPerRack.count(taskId.getRackId()) > perRack && extraCleanedTasks.size() < numActiveRacks / 2) {
-            extraCleanedTasks.add(taskId);
-            LOG.info("Cleaning up task {} to evenly distribute tasks among racks", taskId);
-            taskManager.createTaskCleanup(new SingularityTaskCleanup(pendingRequest.getUser(), TaskCleanupType.REBALANCE_RACKS, now, taskId, Optional.<String>absent(), Optional.<String>absent(), Optional.<SingularityTaskShellCommandRequestId>absent()));
+        rebalanceRacks(request, state, deployStatistics, pendingRequest, maybePendingDeploy, now, remainingActiveTasks);
+      }
+      if (request.getSlaveAttributeMinimums().isPresent()) {
+        rebalanceAttributeDistribution(request, state, deployStatistics, pendingRequest, maybePendingDeploy, now, remainingActiveTasks);
+      }
+    }
+    return numMissingInstances;
+  }
+
+  private void rebalanceAttributeDistribution(
```

Comment:
@ssalinas @baconmania this is a grossly inefficient implementation. Any ideas or suggestions greatly appreciated.

Reply:
My first thought on this is that we'll want to maintain one instance of `extraTasksToClean`, not recreate it each loop. For a single attribute minimum specified, this works just fine. It breaks down more for multiple since we'd be creating multiple cleanups possibly for the same task. Also, you are looping through the tasks twice here which you can likely just do it once.

Current:
 - loop through tasks to find instance attr counts
- loop through minimums and loop through tasks again to determine which to clean

Proposed:
- Loop through tasks to build lists of task ids with/without the attributes you care about (e.g. something like Map<Attr, Map<Boolean, List<TaskId>>> as opposed to just counting.
- Loop through minimums to find what might need a count of cleanups and pop those out of the lists if they aren't already being clean up for a different attribute minimum

There is probably a way we can create less overall cleanups by figuring out which cleanups satisfy the most requirements, but I think that may be over optimizing considering we are designing this currently for a rather specific use case of a single minimum attribute